tests/student/test_analytics.py

```
@@ -0,0 +1,1871 @@
+    with pytest.raises(TypeError):
+        cost.arguments["self"] = TensorRepr((1, 9, 3), torch.float32)  # ty: ignore[invalid-assignment]
+
+    tensors = [TensorRepr((2, 3)), TensorRepr((2, 4))]
+    concatenation = CostRepr("concatenation", torch.ops.aten.cat.default, {"tensors": tensors})
+    tensors.append(TensorRepr((2, 5)))
+
+    assert concatenation.arguments["tensors"] == (TensorRepr((2, 3)), TensorRepr((2, 4)))
+
+
+def test_symbolic_tensor_requires_a_torch_dtype() -> None:
+    with pytest.raises(TypeError, match="torch.dtype or None"):
```

Ruff (0.15.21)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)